File: backend/harness/commands.py
"""
OpenCode-style slash command expansion on the server.

Client handles pure-UI commands (/help, /new, /settings…).
Anything that reaches the backend as a prompt may still start with /plan, /init, etc.
This module normalizes those into agent mode + task text.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class ResolvedCommand:
    task: str
    agent: str  # build | plan
    force_harness: bool
    command: Optional[str] = None


_SLASH = re.compile(r"^/([a-zA-Z0-9_?-]+)(?:\s+([\s\S]*))?$")

# Templates (mirrors frontend + OpenCode initialize/review spirit)
_TEMPLATES: dict[str, tuple[str, str]] = {
    # name -> (agent, template with $ARGUMENTS)
    "init": (
        "plan",
        "Create or update a compact AGENTS.md-style project guide.\n"
        "User focus: $ARGUMENTS\n"
        "Only high-signal, repo-specific guidance.",
    ),
    "review": (
        "plan",
        "You are a code reviewer. Scope: $ARGUMENTS\n"
        "Focus on bugs, edge cases, security, and requirement fit. Be certain.",
    ),
    "test": (
        "build",
        "Write thorough tests for: $ARGUMENTS\n"
        "Implement/fix until tests pass. Use tools and verify with execution.",
    ),
    "fix": (
        "build",
        "Debug and fix:\n$ARGUMENTS\nUse tools, verify, then finish.",
    ),
    "plan": ("plan", "$ARGUMENTS"),
    "build": ("build", "$ARGUMENTS"),
    "code": ("build", "$ARGUMENTS"),
}

_ALIASES = {
    "summarize": "compact",  # client-only usually
    "?": "help",
}


def resolve_slash(text: str, *, default_agent: str = "build") -> ResolvedCommand:
    """
    If `text` starts with a known slash command, expand it.
    Unknown / UI-only commands are left as plain text (client should have handled them).
    """
    t = (text or "").strip()
    m = _SLASH.match(t)
    if not m:
        agent = "plan" if _wants_plan(t) else default_agent
        return ResolvedCommand(task=t, agent=agent, force_harness=False)

    name = m.group(1).lower()
    name = _ALIASES.get(name, name)
    args = (m.group(2) or "").strip()

    if name in _TEMPLATES:
        agent, tmpl = _TEMPLATES[name]
        body = tmpl.replace("$ARGUMENTS", args or "(none)").strip()
        if name in ("plan", "build", "code") and not args:
            # bare /plan or /build — short instruction to switch mode
            body = (
                "Stay in plan mode: analyze only, no file writes."
                if name == "plan"
                else "Use full build tools to implement the next coding request."
            )
        return ResolvedCommand(
            task=body,
            agent=agent,
            force_harness=True,
            command=name,
        )

    # Unknown slash → pass through (may be custom)
    return ResolvedCommand(task=t, agent=default_agent, force_harness=False)


def _wants_plan(task: str) -> bool:
    return bool(
        re.search(
            r"\b(plan only|don't (edit|write)|do not (edit|write)|read only|analyze only)\b",
            (task or "").lower(),
        )
    )

File: frontend/src/lib/commands.ts
/**
 * OpenCode-inspired slash commands for EdgeRunner.
 *
 * Built-ins mirror OpenCode TUI (https://opencode.ai/docs/tui#commands):
 *   /help /new /clear /compact /models /sessions /undo /redo
 *   /export /thinking /details /themes /settings
 * Plus coding agents:
 *   /code /build /plan /agent /init /review
 *
 * Custom templates can be added later via prefs.customCommands.
 */

export type AgentMode = "build" | "plan";

export type CommandKind =
  | "local" // handled entirely in the UI
  | "prompt" // expand to a prompt and send to the agent
  | "force_harness"; // send remaining text with harness forced

export type SlashCommand = {
  name: string;
  aliases?: string[];
  description: string;
  kind: CommandKind;
  /** For kind=prompt: template. $ARGUMENTS = rest of line after command. */
  template?: string;
  /** Preferred agent when this command is used */
  agent?: AgentMode;
  /** Hide from default help (internal) */
  hidden?: boolean;
};

export const BUILTIN_COMMANDS: SlashCommand[] = [
  {
    name: "help",
    aliases: ["?"],
    description: "Show available slash commands",
    kind: "local",
  },
  {
    name: "new",
    aliases: ["clear"],
    description: "Start a new session (clear chat)",
    kind: "local",
  },
  {
    name: "compact",
    aliases: ["summarize"],
    description: "Compact chat context (keep last turns + summary note)",
    kind: "local",
  },
  {
    name: "models",
    aliases: ["model"],
    description: "Open model picker",
    kind: "local",
  },
  {
    name: "settings",
    aliases: ["config", "connect"],
    description: "Open settings drawer",
    kind: "local",
  },
  {
    name: "sessions",
    aliases: ["resume", "continue"],
    description: "Show session / connection status",
    kind: "local",
  },
  {
    name: "export",
    description: "Export conversation as Markdown (download)",
    kind: "local",
  },
  {
    name: "undo",
    description: "Remove last user+assistant turn",
    kind: "local",
  },
  {
    name: "redo",
    description: "Restore last undone turn",
    kind: "local",
  },
  {
    name: "thinking",
    description: "Toggle visibility of agent traces / thoughts",
    kind: "local",
  },
  {
    name: "details",
    description: "Toggle tool-detail density in CLI view",
    kind: "local",
  },
  {
    name: "cli",
    description: "Switch to CLI (terminal) view",
    kind: "local",
  },
  {
    name: "chat",
    description: "Switch to chat (markdown) view",
    kind: "local",
  },
  {
    name: "view",
    description: "Toggle CLI ↔ chat view",
    kind: "local",
  },
  {
    name: "agent",
    description: "Show or set agent: /agent build | /agent plan",
    kind: "local",
  },
  {
    name: "engine",
    description: "Switch agent engine: /engine hermes | native",
    kind: "local",
  },
  {
    name: "loop",
    description: "Run a task until done: /loop [n] <task> (default 3 rounds)",
    kind: "local",
  },
  {
    name: "retry",
    description: "Resend the last user message",
    kind: "local",
  },
  {
    name: "stop",
    aliases: ["abort", "cancel"],
    description: "Cancel the in-flight run (server-side too)",
    kind: "local",
  },
  {
    name: "memory",
    description: "Persistent notes: /memory add <text> | list | clear",
    kind: "local",
  },
  {
    name: "system",
    description: "Custom system prompt: /system <text> | show | clear",
    kind: "local",
  },
  {
    name: "build",
    description: "Switch to Build agent (full tools) and optional task",
    kind: "force_harness",
    agent: "build",
  },
  {
    name: "plan",
    description: "Switch to Plan agent (readonly) and optional task",
    kind: "force_harness",
    agent: "plan",
  },
  {
    name: "code",
    description: "Force coding harness for the rest of the message",
    kind: "force_harness",
    agent: "build",
  },
  {
    name: "init",
    description: "Guided AGENTS.md / project setup (OpenCode /init)",
    kind: "prompt",
    agent: "plan",
    template: `Create or update a compact AGENTS.md (or project guidance) for the current coding environment.

User focus (if any): $ARGUMENTS

Include only high-signal, repo-specific guidance an agent would miss:
- exact run/test/lint commands
- architecture entrypoints
- constraints and gotchas
Omit generic advice. Prefer short bullets.`,
  },
  {
    name: "review",
    description: "Review recent changes [commit|branch|pr] (OpenCode /review)",
    kind: "prompt",
    agent: "plan",
    template: `You are a code reviewer. Review the requested scope and provide actionable feedback.

Input: $ARGUMENTS
(If empty, review the most recent solution / conversation code.)

Focus on:
1. Bugs and logic errors
2. Edge cases and error handling
3. Security issues
4. Fit with stated requirements

Be certain before flagging. Diffs alone are not enough — reason about intended behavior.
Do not invent hypothetical problems.`,
  },
  {
    name: "test",
    description: "Ask agent to write/run thorough tests for the task",
    kind: "prompt",
    agent: "build",
    template: `Write thorough tests for: $ARGUMENTS

Then implement or fix the solution until tests pass. Use the coding harness tools.
Prefer assert-style unit tests. Report final code and test results.`,
  },
  {
    name: "fix",
    description: "Debug and fix based on error context",
    kind: "prompt",
    agent: "build",
    template: `Debug and fix the following issue. Reproduce with tests if possible, then patch.

$ARGUMENTS

Use tools, verify with execution, call done when fixed.`,
  },
];

const byName = new Map<string, SlashCommand>();
for (const c of BUILTIN_COMMANDS) {
  byName.set(c.name, c);
  for (const a of c.aliases || []) byName.set(a, c);
}

export function listCommands(): SlashCommand[] {
  return BUILTIN_COMMANDS.filter((c) => !c.hidden);
}

export function findCommand(name: string): SlashCommand | undefined {
  return byName.get((name || "").toLowerCase().replace(/^\//, ""));
}

export type ParsedSlash =
  | { kind: "none"; text: string }
  | {
      kind: "unknown";
      name: string;
      raw: string;
      suggestions: SlashCommand[];
    }
  | {
      kind: "command";
      command: SlashCommand;
      args: string;
      raw: string;
    };

/** Parse a leading /command from the input line. */
export function parseSlash(input: string): ParsedSlash {
  const text = (input || "").trim();
  if (!text.startsWith("/")) return { kind: "none", text: input };

  // /command rest…
  const m = text.match(/^\/([a-zA-Z0-9_?-]+)(?:\s+([\s\S]*))?$/);
  if (!m) {
    // "/", "/ …" — malformed command, never send to the model
    return { kind: "unknown", name: text.slice(1), raw: text, suggestions: [] };
  }
  const name = m[1].toLowerCase();
  const args = (m[2] || "").trim();
  const command = findCommand(name);
  if (!command) {
    // Typo / partial like "/sett" — hint locally instead of sending to the model
    return {
      kind: "unknown",
      name,
      raw: text,
      suggestions: filterCommands(name).slice(0, 3),
    };
  }
  return { kind: "command", command, args, raw: text };
}

export function expandTemplate(template: string, args: string): string {
  return template
    .replace(/\$ARGUMENTS/g, args || "(none)")
    .replace(/\$0/g, args || "")
    .trim();
}

export function helpText(): string {
  const lines = [
    "EdgeRunner slash commands (OpenCode-inspired)",
    "",
    "Usage: /command [arguments]",
    "",
  ];
  const seen = new Set<string>();
  for (const c of listCommands()) {
    if (seen.has(c.name)) continue;
    seen.add(c.name);
    const alias =
      c.aliases && c.aliases.length
        ? ` (aliases: ${c.aliases.map((a) => "/" + a).join(", ")})`
        : "";
    lines.push(`  /${c.name.padEnd(12)} ${c.description}${alias}`);
  }
  lines.push("");
  lines.push("Also:");
  lines.push("  Tab in composer is unused — use /agent build|plan");
  lines.push("  Prefix ! is reserved for future local shell (OpenCode-style)");
  lines.push("  Coding tasks auto-route to the OpenCode-style tool harness");
  return lines.join("\n");
}

/** Fuzzy filter for command palette while typing `/…` */
export function filterCommands(query: string): SlashCommand[] {
  const q = (query || "").toLowerCase().replace(/^\//, "");
  if (!q) return listCommands();
  return listCommands().filter(
    (c) =>
      c.name.startsWith(q) ||
      c.aliases?.some((a) => a.startsWith(q)) ||
      c.description.toLowerCase().includes(q)
  );
}
